File: android/app/src/main/java/es/manabe/yomiyasu/core/readers/ImageFolderPages.kt
package es.manabe.yomiyasu.core.readers

import es.manabe.yomiyasu.core.mokuro.MokuroBook
import es.manabe.yomiyasu.core.mokuro.MokuroPage
import es.manabe.yomiyasu.core.mokuro.MokuroSize

/**
 * Construye las páginas de un tomo sin mokuro (carpeta de imágenes) a partir
 * de la lista de imágenes que devuelve el backend. Se reutiliza [MokuroPage]
 * sin cajas de texto: los spreads, el progreso y el pager funcionan igual que
 * con un tomo de mokuro y la vista usa [ImagePageView].
 */
object ImageFolderPages {
    /**
     * Ruta de una página relativa a la carpeta de la serie. En mokuro el
     * `imagePath` ya incluye la carpeta de imágenes (sale del html); en los
     * tomos de imágenes hay que prefijarla para que el lector y las descargas
     * usen la misma ruta (`mangas/{serie}/{imagesFolder}/{archivo}`).
     */
    fun joinedPath(imagesFolder: String?, fileName: String): String =
        if (imagesFolder.isNullOrEmpty()) fileName else "$imagesFolder/$fileName"

    fun makeBook(pagePaths: List<String>, imagesFolder: String?): MokuroBook =
        MokuroBook(
            pages = pagePaths.mapIndexed { index, path ->
                MokuroPage(
                    id = index,
                    imagePath = joinedPath(imagesFolder, path),
                    size = MokuroSize(0f, 0f),
                    boxes = emptyList(),
                )
            },
        )
}

File: ios/Yomiyasu/Core/Readers/ImageFolderPages.swift
import CoreGraphics

/// Construye las páginas de un tomo sin mokuro (carpeta de imágenes) a partir
/// de la lista de imágenes que devuelve el backend. Se reutiliza `MokuroBook`
/// sin cajas de texto: el resto del lector (spreads, zoom, progreso) funciona
/// igual que con un tomo de mokuro.
enum ImageFolderPages {
    /// Proporción de reserva (página de manga) hasta conocer el tamaño real.
    static let defaultPageSize = CGSize(width: 1400, height: 2000)

    /// Ruta de una página relativa a la carpeta de la serie. En mokuro el
    /// `imagePath` ya incluye la carpeta de imágenes (sale del html); en los
    /// tomos de imágenes hay que prefijarla para que el lector y las descargas
    /// usen la misma ruta (`mangas/{serie}/{imagesFolder}/{archivo}`).
    static func joinedPath(imagesFolder: String?, fileName: String) -> String {
        guard let imagesFolder, !imagesFolder.isEmpty else { return fileName }

        return "\(imagesFolder)/\(fileName)"
    }

    static func makeBook(pagePaths: [String], imagesFolder: String?, firstPageSize: CGSize?) -> MokuroBook {
        let size = firstPageSize ?? defaultPageSize

        let pages = pagePaths.enumerated().map { index, path in
            MokuroPage(
                id: index,
                imagePath: joinedPath(imagesFolder: imagesFolder, fileName: path),
                size: size,
                boxes: []
            )
        }

        return MokuroBook(pages: pages)
    }
}
